### PR TITLE
fix: Allows role_arn to be able to pass from module input

### PR DESCRIPTION
**Pull request by Bharathkumarraju**
## Description

This PR introduces the ability to pass an existing role_arn from an external IAM module to the eks-pod-identity module. Instead of always creating a new IAM role within the eks-pod-identity module, users can now provide an externally created role_arn for associating with the EKS service account. This enhancement adds more flexibility, especially for users who need to reuse existing IAM roles for compliance or governance reasons.

Changes made:
Updated the module to accept role_arn as an input in the associations map.
Modified the aws_eks_pod_identity_association resource to use the provided role_arn when supplied.
Retained backward compatibility by keeping the option to create a new IAM role when role_arn is not provided and create = true.

## Motivation and Context

This change is required for cases where users want to reuse an existing IAM role rather than creating a new one inside the eks-pod-identity module. This is particularly useful for scenarios where roles are centrally managed and need to follow predefined policies, tags, or naming conventions.

Without this change, the module would always create a new IAM role, even when an existing role is more appropriate. By passing the role_arn, it provides users with more flexibility and control over the roles used by their EKS workloads.

## Breaking Changes
There are no breaking changes.

Backward compatibility is preserved by keeping the default behavior, which is to create a new IAM role if role_arn is not provided and `create = true.`

## How Has This Been Tested?
```
module "pod_identity" {
  source  = "terraform-aws-modules/eks-pod-identity/aws"
  version = "~> 1.5.0"
  create = false  # Disable IAM role creation
  associations = {
    example_pod_identity = {
      cluster_name    = "example-eks-cluster"
      namespace       = "example"
      service_account = "example-test"
      role_arn        = module.example_role.arn # Pass the existing role_arn
      tags = {
        env     = "test"
        purpose = "Allow example-app to use pod identity"
      }
    }
  }
}
```

### CODE REVIEW

**Comment by bryantbiggs** [comment]
Why?

**Comment by Bharathkumarraju** [comment]
> Why?

@bryantbiggs we are actually migrating from IRSA to Pod-identity we are trying to use existing roles with this module i found currently this module doesn't support it ... so tried to update it if someone also looking for it. Actually i have locally updated this module tested against our existing role

**Comment by Bharathkumarraju** [comment]
Also if I don't use `create = false` in the module and tried using only `role_arm` it still created new role something like this `arn:aws:iam::12345678912:role/-20241009022917608700000001`

```
}
bharathkumardasaraju@404-test-html$ aws eks describe-pod-identity-association --cluster-name example-eks --association-id a-z1f43ezn4ealpo0pa
{
    "association": {
        "clusterName": "example-eks",
        "namespace": "example",
        "serviceAccount": "example-test",
        "roleArn": "arn:aws:iam::12345678912:role/-20241009022917608700000001",
        "associationArn": "arn:aws:eks:ap-southeast-1:12345678912:podidentityassociation/example-eks/a-z1f43ezn4ealpo0pa",
        "associationId": "a-z1f43ezn4ealpo0pa",
        "tags": {
            "env": "test",
            "purpose": "Allow test-app to use pod identity"
        },
        "createdAt": "2024-10-09T16:58:34.684000+08:00",
        "modifiedAt": "2024-10-09T16:58:34.684000+08:00"
    }
}

```

**Comment by bryantbiggs** [comment]
you don't need to migrate the roles - to migrate from IRSA to pod identity would look something like:
1. Ensure you have the pod identity agent deployed and the pods that will be using pod identity are using an AWS SDK version that support pod identity
2. Create the pod identity role with policy and pod identity association - pod identity now exists but the pods are not using it
3. You now have the option to either just roll the deployment so that new pods come up and the AWS credential chain should then use pod identity, or you can update your manifests to remove the annotation which would therefore also roll the deployment and ensure new pods come up to pick up the right creds from the AWS credential chain

however, I don't think it would make sense for this module to accept an IAM role from the outside. its a very thin layer of functionality and removing the role creation would make it rather pointless

**Comment by Bharathkumarraju** [comment]
okay thanks got it @bryantbiggs  we are not removing the role_creation though its flag right :)

**Comment by bryantbiggs** [comment]
but why would we create a role if we aren't using it?

**Comment by Bharathkumarraju** [comment]
Thanks for the clarification, @bryantbiggs 

I see your point about the role creation logic. The intention here is not to remove the default role creation but to provide an option for users who already have predefined IAM roles managed outside of this module, especially in environments with strict governance or compliance policies.

The use case we are addressing is where roles are centrally managed, often for security or audit reasons, and organizations may have existing policies or conventions they need to follow. This enhancement allows the flexibility to reuse those roles while retaining the module's existing behavior of creating a new role when none is provided.

This would allow users who want to manage IAM roles externally to have more control while keeping the module useful for those who don't need such control. So, if role_arn is supplied, we leverage that; otherwise, the module creates a new role.

In short, it offers more flexibility without forcing a major change to the current design.

Would love to hear your thoughts on this approach!

**Comment by bryantbiggs** [comment]
That doesn't make any sense - how would this work if you aren't allowed to create a role?

**Comment by Bharathkumarraju** [comment]
Thanks again for your feedback @bryantbiggs 

I think we might be talking about slightly different scenarios here. Let me clarify:

When  [using this](https://github.com/terraform-aws-modules/terraform-aws-eks-pod-identity/blob/master/variables.tf#L4) `create = true (the default)`,  the module creates a new IAM role. This works perfectly for users who don't need to manage IAM roles externally.
When `create = false`, as in our use case, we want to avoid creating a new role because our IAM roles are centrally managed and predefined for compliance reasons. In this case, we pass the role_arn from an external IAM module.

The goal here is not to change the default behavior but to provide flexibility for users who already have existing roles. This would prevent creating new IAM roles unnecessarily while allowing organizations with stricter role management policies to use the module.

So, in short, the module retains the ability to create new roles when needed, but users who don't want that (like in our case) can set `create = false` and provide the `role_arn`

**Comment by Bharathkumarraju** [comment]
@bryantbiggs looks we may need two changes here

in the variables.tf is this
```
variable "role_arn" {
  description = "The IAM role ARN to associate with the Pod Identity, if using an existing role"
  type        = string
  default     = null
}
```

and the main.tf

```
resource "aws_eks_pod_identity_association" "this" {
  for_each = { for k, v in var.associations : k => v }

  cluster_name    = try(each.value.cluster_name, var.association_defaults.cluster_name)
  namespace       = try(each.value.namespace, var.association_defaults.namespace)
  service_account = try(each.value.service_account, var.association_defaults.service_account)

  # Use the external role_arn if provided, otherwise fall back to the internally created role
  role_arn = coalesce(try(each.value.role_arn, var.role_arn), aws_iam_role.this[0].arn)

  tags = merge(var.tags, try(each.value.tags, var.association_defaults.tags, {}))
}
```

**Comment by Bharathkumarraju** [comment]
And then my external module looks like this 

```
module "pod_identity" {
  source  = "terraform-aws-modules/eks-pod-identity/aws"
  version = "~> 1.5.0"
  create  = false  # Disable IAM role creation

  associations = {
    example_pod_identity = {
      cluster_name    = "example-eks-cluster"
      namespace       = "example"
      service_account = "example-test"
      role_arn        = module.example_role.arn  # Provide the external role
      tags = {
        env     = "test"
        purpose = "Allow example-app to use pod identity"
      }
    }
  }
}
```

**Comment by bryantbiggs** [comment]
that still doesn't make any sense - you are only using the pod identity association resource which you would be better of just using that resource directly - not through the module
